R/xml.R

```diff
@@ -0,0 +1,183 @@
+        cat(table$header, file = conv_pipe, sep = "\n")
+        cat("", file = conv_pipe, sep = "\n")
+        utils::write.table(table$data, file = conv_pipe, sep=",", row.names = FALSE, col.names = TRUE, quote = FALSE)
+        cat("", file = conv_pipe, sep = "\n")
+      }
+    }
+
+    dot
+  }
+}
+
+#' Run the CSV to XML conversion using the model interface tool.
+#'
+#' This method should be the final call in a pipeline started with \code{\link{create_xml}}
+#' and one or more calls to \code{\link{add_xml_data}}.
+#' Please see the wiki for more details \url{https://github.com/JGCRI/gcamdata/wiki/XML-Conversion}
+#'
+#' Note that this method relies on Java to run the conversion.  To avoid errors for
+#' users who do not have Java installed it will check the global option
+#' \code{gcamdata.use_java} before attempting to run the conversion.  If the flag
+#' is set to \code{FALSE} a warning will be issued and the conversion skipped.
+#' To enable the use of Java a user can set \code{options(gcamdata.use_java=TRUE)}
+#'
+#' @param dot The current state of the pipeline started from \code{create_xml}
+#' @return The argument passed in unmodified in case a user wanted run the
+#' conversion again at a later time.
+#' @author PP March 2017
+#' @export
+run_xml_conversion <- make_run_xml_conversion()
+
+#' Add a table to an XML pipeline that instructs the ModelInterface to rename
+#' LandNodeX to LandNode.
+#'
+#' Such a table is necessary to help work around limitations in the XML processing
+#' that node names of the same name can not be nested with in each other:
+#' LandNode/LandNode thus instead we say LandNode1/LandNode2 and rename as the last step.
+#' Therefore in most cases a user should add this table near the end of the XML pipeline.
+#' @param dot The current state of the pipeline started from \code{create_xml}
+#' @return A "data structure" to hold the various parts needed to run the model
+#' interface CSV to XML conversion.
+#' @author Pralit Patel
+#' @export
+add_rename_landnode_xml <- function(dot) {
+  land_name_table <- tibble(from=paste0("LandNode", seq(1,5)),to="LandNode")
+
+  add_xml_data(dot, land_name_table, "NodeRename", NULL)
+}
+
+#' add_node_equiv_xml
+#'
+#' Add a table to an XML pipeline that instructs the ModelInterface to treat
+#' tags in the same class as the same when enabled.  Thus not requiring multiple
+#' headers for instance to read in a share-weight into a technology
+#' intermittent-technology or tranTechnology.  A user taking advantage of this
+#' feature would then read this table early in the XML pipeline.
+#' @param dot The current state of the pipeline started from \code{create_xml}
+#' @param equiv_class A name of the equivalence class to add.  This could be any
+#' key in the list \code{XML_NODE_EQUIV}.
+#' @return A "data structure" to hold the various parts needed to run the model
+#' interface CSV to XML conversion.
+#' @author Pralit Patel
+#' @export
```

Comment:
Yes, I foresee users will need to make "one off" XML files from time most frequently to add some policy lever into some technology.  They could grab the tables necessary to define the structure out of the results from gcamdata, add in their levers,  and use these XML methods to create the XML input file and write it in some R script so it is reproducible.
Although I realize this is likely just pure fantasy and what really happens is they take the CSV results from gcamdata, load it into excel and do the step manually to create the XML files in which case exporting the functions isn't necessary.
I'd prefer to at least make it as easy as possible to do it in a reproducible way.